kernel/sched.c

```diff
@@ -400,7 +400,7 @@ static void create_kernel_process(void)
       sizeof(struct process) + sizeof(struct task)
    ] ALIGNED_AT(sizeof(void *));
 
-   struct task *s_kernel_ti = (struct task *)kernel_proc_buf;
+   struct task *s_kernel_ti = (struct task *)(void *)kernel_proc_buf;
```

Comment:
Here the kernel_proc_buf should have a different alignment, not variable below to use a different cast. Can you try with:
```
   static char kernel_proc_buf[
      sizeof(struct process) + sizeof(struct task)
   ] ALIGNED_AT(alignof(struct process));
```
If that still doesn't work, you can keep just the `void *` cast and add two static asserts:
```
STATIC_ASSERT(alignof(struct task) == alignof(kernel_proc_buf));
STATIC_ASSERT(alignof(struct process) == alignof(kernel_proc_buf));
```

Reply:
But I start to sense the problem. Even if the alignment is numerically correct, the compiler checks the symbolic value of the alignment and that is different.

Reply:
> Here the `kernel_proc_buf` variable should have a different alignment, not the variable below to use a different cast. Can you try with:
> 
> ```
>    static char kernel_proc_buf[
>       sizeof(struct process) + sizeof(struct task)
>    ] ALIGNED_AT(alignof(struct task));
> ```
> 
> If that still doesn't work, you can keep just the `void *` cast and add two static asserts:

It doesn't work, I'll just have to add these two static asserts

Reply:
An error occurred in clang，how should we fix it?

```
/__w/1/s/kernel/sched.c:403:42: error: '_Alignof' applied to an expression is a GNU extension [-Werror,-Wgnu-alignof-expression]
   STATIC_ASSERT(alignof(struct task) == alignof(kernel_proc_buf));
                                         ^
/usr/lib/llvm-14/lib/clang/14.0.0/include/stdalign.h:15:17: note: expanded from macro 'alignof'
#define alignof _Alignof
                ^
```

Reply:
@aenrbes I was just writing you about that. I think that the right solution is use use `_Alignas`, which is available since C11: https://en.cppreference.com/w/c/language/_Alignas
Just add an `ALIGNAS()` macro in basic_defs.h for that, as in C23 they renamed it to `alignas()`.

If you wanna test clang builds of ARCH=i386 locally, you can run the `cmake_run` script using the `--contrib` flag. It will use `clang` and a bunch of other contributor-only flags like `WCONV=1` which turns on `-Wconversion` on `clang`.

Reply:
Do you mean to replace `ALIGNED_AT` with `ALIGNAS` and require `kernel_proc_buf` to be aligned to the `struct task`, and does not require `STATIC_ASSERT`, as in the code snippet below? 
From what I understand about it, there is no way in standard C to get the alignment requirements of an expression, as is the case with alignof(expression)

```
   static char ALIGNAS(struct task) kernel_proc_buf[
      sizeof(struct process) + sizeof(struct task)
   ];

   struct task *s_kernel_ti = (void *)kernel_proc_buf;
   struct process *s_kernel_pi = (struct process *)(s_kernel_ti + 1);
```

Reply:
@aenrbes Yeah, I meant exactly that. If that still doesn't work, let's hack it with `void *` (whatever you did before) plus the ASSERTs. If your new update passes all the pre-existing builds, I'll just merge it.

As a follow-up I'll make a change to make all of that super compliant: I'll make a new struct type like `struct task_and_process` with two explicit fields and use that here and in elsewhere. That will also completely eliminate expressions like `(struct process *)(s_kernel_ti + 1)`. Not sure why I didn't want to do it the super-clear way originally. I guess that it felt a bit cooler to not declare a new struct.